### PR TITLE
Added prometheus metrics and endpoint.

### PR DESCRIPTION
**Pull request by kousik93**
Added prometheus metrics and endpoint.

### CODE REVIEW

File: cmd/kapprover/main.go [hunk 1]
```diff
@@ -46,9 +46,21 @@ func main() {
 		return
 	}
 
+	go serveMetrics()
 	kapprover.HandleRequests(filters, deniers, warners, *deleteAfter, client)
 }
 
+func serveMetrics() {
```
Comment (johngmyers) [inline cmd/kapprover/main.go]: This should be put into the kapprover package so it is usable by people who need to add in their own inspectors.

File: cmd/kapprover/main.go [hunk 2]
```diff
@@ -46,9 +46,21 @@ func main() {
...
+	http.HandleFunc("/healthz", func(w http.ResponseWriter, _ *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte("OK"))
+	})
+
+	http.Handle("/metrics", promhttp.Handler())
+
+	log.Fatal(http.ListenAndServe(":9201", nil))
```
Comment (johngmyers) [inline cmd/kapprover/main.go]: 9201 is the portfor "Remote storage bridge example code". We should probably use something like 8081.

The port number should probably be passed in from main, so people using the kapprover package from their own main can easily change it.

File: kapprover/kapprover.go [hunk 3]
```diff
@@ -14,14 +11,66 @@ import (
...
 
 var (
 	scheduled    = map[string]bool{}
 	scheduledMux sync.Mutex
 )
 
+var (
+	requestsApproved = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "ieng_kapprover_requests_approved",
```
Comment (johngmyers) [inline kapprover/kapprover.go]: remove "ieng_" prefix.

File: kapprover/kapprover.go [hunk 4]
```diff
@@ -14,14 +11,66 @@ import (
...
+		prometheus.CounterOpts{
+			Name: "ieng_kapprover_requests_filtered",
+			Help: "Number of filtered requests.",
+		},
+		[]string{},
+	)
+	requestsError = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "ieng_kapprover_requests_error",
+			Help: "Number of requests that errored out.",
```
Comment (johngmyers) [inline kapprover/kapprover.go]: Number of requests encountering an error

File: kapprover/kapprover.go [hunk 5]
```diff
@@ -73,10 +122,12 @@ func tryApprove(filters inspectors.Inspectors, deniers inspectors.Inspectors, wa
 		for _, filter := range filters {
 			message, err := filter.Inspector.Inspect(client, request)
 			if err != nil {
+				requestsError.WithLabelValues("filterError").Inc()
 				return err
 			}
 			if message != "" {
 				log.Infof("Skipping %q from %q: %s", request.Name, request.Spec.Username, message)
+				requestsFiltered.WithLabelValues().Inc()
```
Comment (johngmyers) [inline kapprover/kapprover.go]: Throw in the insepctor.Name as a label value?

File: kapprover/kapprover.go [hunk 6]
```diff
@@ -73,10 +122,12 @@ func tryApprove(filters inspectors.Inspectors, deniers inspectors.Inspectors, wa
 		for _, filter := range filters {
 			message, err := filter.Inspector.Inspect(client, request)
 			if err != nil {
+				requestsError.WithLabelValues("filterError").Inc()
```
Comment (johngmyers) [inline kapprover/kapprover.go]: Use the inspector.Name as the label value?

File: kapprover/kapprover.go [hunk 7]
```diff
@@ -90,12 +141,14 @@ func tryApprove(filters inspectors.Inspectors, deniers inspectors.Inspectors, wa
 		for _, denier := range deniers {
 			message, err := denier.Inspector.Inspect(client, request)
 			if err != nil {
+				requestsError.WithLabelValues("denyError").Inc()
```
Comment (johngmyers) [inline kapprover/kapprover.go]: Use  the denier.Name as the label value?

File: kapprover/kapprover.go [hunk 8]
```diff
@@ -105,6 +158,7 @@ func tryApprove(filters inspectors.Inspectors, deniers inspectors.Inspectors, wa
 				message, _ := warner.Inspector.Inspect(client, request)
 				if message != "" {
 					log.Warnf("Approving CSR %q from %q despite %s: %s", request.Name, request.Spec.Username, warner.Name, message)
+					requestsWarned.WithLabelValues(warner.Name).Inc()
```
Comment (johngmyers) [inline kapprover/kapprover.go:175]: There can be multiple warnings per approved CSR, so we should change the metric help to "Number of warnings on approved requests".

File: kapprover/kapprover.go [hunk 9]
```diff
@@ -123,7 +177,7 @@ func tryApprove(filters inspectors.Inspectors, deniers inspectors.Inspectors, wa
 				}
 				continue
 			}
-
+			requestsError.WithLabelValues("csrSubmitError").Inc()
```
Comment (johngmyers) [inline kapprover/kapprover.go]: More like "updateApproval".

**Review by johngmyers — CHANGES_REQUESTED** [review]
Appears to be missing changes to Gopkg.toml

File: kapprover/kapprover.go (hunk 7 above)
Comment (kousik93) [reply kapprover/kapprover.go]: Done.

File: kapprover/kapprover.go (hunk 6 above)
Comment (kousik93) [reply kapprover/kapprover.go]: Done

File: cmd/kapprover/main.go (hunk 1 above)
Comment (kousik93) [reply cmd/kapprover/main.go]: Moved

File: cmd/kapprover/main.go (hunk 2 above)
Comment (kousik93) [reply cmd/kapprover/main.go]: Fixed

File: kapprover/kapprover.go (hunk 3 above)
Comment (kousik93) [reply kapprover/kapprover.go]: Done

File: kapprover/kapprover.go (hunk 4 above)
Comment (kousik93) [reply kapprover/kapprover.go]: Done

File: kapprover/kapprover.go (hunk 5 above)
Comment (kousik93) [reply kapprover/kapprover.go]: Done

File: kapprover/kapprover.go (hunk 8 above)
Comment (kousik93) [reply kapprover/kapprover.go:175]: Done

File: kapprover/kapprover.go (hunk 9 above)
Comment (kousik93) [reply kapprover/kapprover.go]: Changed.

File: kapprover/kapprover.go [hunk 10]
```diff
@@ -14,13 +12,80 @@ import (
...
+			Help: "Number of filtered requests.",
+		},
+		[]string{"reason"},
+	)
+	requestsError = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "kapprover_requests_error",
+			Help: "Number of requests encountering an error.",
+		},
+		[]string{"message"},
```
Comment (johngmyers) [inline kapprover/kapprover.go]: reason?

**Review by johngmyers — CHANGES_REQUESTED** [review]
My comment was to move the registerPrometheusMetrics() back to HandleRequests. The serveMetrics() code (or ServePrometheusMetrics()) should be public in the kapprover package

File: kapprover/kapprover.go [hunk 11]
```diff
@@ -14,14 +12,79 @@ import (
...
+		w.Write([]byte("OK"))
+	})
+
+	http.Handle("/metrics", promhttp.Handler())
+
+	log.Fatal(http.ListenAndServe(":"+strconv.Itoa(port), nil))
+}
+
 func HandleRequests(filters inspectors.Inspectors, deniers inspectors.Inspectors, warners inspectors.Inspectors, deleteAfter time.Duration, client *kubernetes.Clientset) {
+	//Register prometheus metrics
```
Comment (johngmyers) [inline kapprover/kapprover.go:85]: nit: comment is redundant

**Review by johngmyers — APPROVED** [review]

**Comment by johngmyers** [comment]
Squashed and merged

**Pull request closed without merging**